### ISSUE TITLE
Can't freeze Sketchup::Color

### ISSUE THREAD

**Issue by linying1991** [SketchUp/api-issue-tracker#826]
# Bug Reports

1. SketchUp/LayOut Version: 2022
2. OS Platform: Windows 11

## Sketchup::Color can't freeze
The correct result should be to throw an exception
![1](https://user-images.githubusercontent.com/38310238/185524401-ef354e54-fe66-4202-9e89-0af23c2351ac.png)

## String can freeze
![2](https://user-images.githubusercontent.com/38310238/185524405-3c53af2c-6b17-4b19-a879-2f425e2bca1f.png)

**Comment by thomthom** [comment]
hm.. I don't think anything in the SU API checks the frozen flag...

**Comment by DanRathbun** [comment]
**I do not agree that this is a bug** It is an implementation oversight (like how `clone` does not work for many API objects.)

The main difference here is that [`Geom`](https://ruby.sketchup.com/Geom.html) classes are virtual geometric helper classes that do not exist in the model database.

A color object **is** part of a material object held in the model's collection data. Having it frozen (freezable) makes **no good sense**. 

**What happens when a user manually edits a material in the GUI using the frozen color object?**
Should a dialog pop up saying "Sorry this color is frozen! Move along to something else ..." ? (No of course not.)

Objects held by the model and editable by the GUI, should have their [`#freeze`](https://ruby-doc.org/core-2.7.2/Object.html#method-i-freeze), [`#taint`](https://ruby-doc.org/core-2.7.2/Object.html#method-i-taint) and [`#untrust`](https://ruby-doc.org/core-2.7.2/Object.html#method-i-untrust) methods [undefined](https://ruby-doc.org/core-2.7.2/Module.html#method-i-undef_method).

#### Entity and it's subclass Drawingelement
(I think we talked about this back in the v7 or v8 days on SketchUcation.)
```ruby
class Entity

  undef_method(:freeze)

  undef_method(:taint)

  # Scheduled to be removed in Ruby 3.2:
  undef_method(:untrust) if Object.public_methods.include?(:untrust)

end
```
#### Collection and some other classes that are direct [`Object`](https://ruby-doc.org/core-2.7.2/Object.html) subclasses should also not be freezable:

```ruby
[
  # None of these classes should be freezable, etc.:
  Sketchup::Axes,
  Sketchup::Classifications,
  Sketchup::Color,
  Sketchup::Console,
  Sketchup::Entities,
  Sketchup::EntitiesBuilder,
  Sketchup::ExtensionsManager,
  Sketchup::ImageRep,
  Sketchup::InputPoint,
  Sketchup::InstancePath,
  Sketchup::Menu,
  Sketchup::Model,
  Sketchup::PickHelper,
  Sketchup::Selection,
  Sketchup::Set,
  Sketchup::TextureWriter,
  Sketchup::Tools,
  Sketchup::UVHelper,
  Sketchup::View

].each do |klass|
  klass.class_eval {
    undef_method(:freeze)
    undef_method(:taint)
    # Scheduled to be removed in Ruby 3.2:
    undef_method(:untrust) if ::Object.public_methods.include?(:untrust)
  }
end
```

#### UI classes

It may be that coders may wish to have their [`Command`](https://ruby.sketchup.com/UI/Command.html), [`Toolbar`](https://ruby.sketchup.com/UI/Toolbar.html), etc. objects frozen to prevent nefarious manipulation. But this might also prevent the status bar text property from being updated.

Freezing a [`HtmlDialog`](https://ruby.sketchup.com/UI/HtmlDialog.html) might prevent the callbacks from being reattached if it is reopened, so freezing a dialog is likely **not** desirable in most cases.

----

After undefining attempts to freeze somethin looks like this"
```ruby
Sketchup.active_model.materials.freeze
#=> Error: #<NoMethodError: undefined method `freeze' for #<Sketchup::Materials:0x0000029699df30b8>>
```
~

**Comment by thomthom** [comment]
> What happens when a user manually edits a material in the GUI using the frozen color object?

The SketchUp API returns a copy of the color object. All `Sketchup::Color` objects are standalone, owned by the Ruby object itself. Color, Point3d/2d, Vector3d/2d - they are all plain data objects that doesn't share ownership.

For Entity objects et. al. you're right, freezing them would not work well.

**Comment by DanRathbun** [comment]
> All `Sketchup::Color` objects are standalone, owned by the Ruby object itself.

Do you mean they are just transient data objects as well .. ie, Ruby wrapped up C++ data separate from the C++ Material objects ?

**Comment by thomthom** [comment]
Yes. Whenever you get a point/vector or color you are getting a copy of the value. Thus, if you access the same `vertex.position` a lot, caching might yield performance boost.

**Comment by DanRathbun** [comment]
Okay thanks. I still do not see much good reason to freeze an extension's colors, points or vectors as the extension is in control of it's own objects. It is not likely that any other extension is going to try to modify another extension's objects.

**Comment by thomthom** [comment]
I find it useful to freeze objects used for constants, because it's easy inadvertently pass that object along in your code flow and accidentally mutate it. Freezing it ensures you get an error instead of silently carrying on.

**Comment by MSP-Greg** [comment]
> I find it useful to freeze objects used for constants

There are many reasons to freeze objects.  I've seen frozen objects that are passed to yield blocks, where the object is passed for its data, but it shouldn't be modified.  Obviously, there are many ways to do things in Ruby...

Ideally, the API should freeze them or raise an error if they can't be.
